### PR TITLE
BUG: fix bug where appending unordered ``CategoricalIndex`` variables overrides index

### LINKED ISSUE #24845: Concatenating unordered CategoricalIndex overrides indices
A very similar bug to #20425, but with `CategoricalIndex` instead of `Categorical`:

#### Code Sample

```python
cat1 = pd.CategoricalIndex(
    ["a", "a",],
    categories=["a", "b"],
    ordered=False)
cat2 = pd.CategoricalIndex(
    ["b", "b"],
    categories=["b", "a"],
    ordered=False)

df1 = pd.DataFrame({"A": [1, 2]}, index=cat1)
df2 = pd.DataFrame({"A": [3, 4]}, index=cat2)
print(pd.concat((df1, df2)))  # cat2 disappeared from index
#    A
# a  1
# a  2
# a  3
# a  4
```
#### Problem description

When the indices are unordered, a concatenation operation may yield wrong indices for the resulting DataFrame. If `cat2`'s categories were `["a", "b"]` the concatenation operation would've been successful.

#### Expected Output
```python
cat3 = pd.CategoricalIndex(
    ["a", "a", "b", "b"], 
    categories=["a", "b"], 
    ordered=False)
print(pd.DataFrame({"A": [1, 2, 3, 4]}, index=cat3))
#    A
# a  1
# a  2
# b  3
# b  4
```
#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.7.final.0
python-bits: 64
OS: Linux
OS-release: 4.18.0-13-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.23.4
pytest: 3.8.1
pip: 18.1
setuptools: 40.2.0
Cython: 0.28.5
numpy: 1.15.4
scipy: 1.2.0
pyarrow: None
xarray: 0.11.0
IPython: 7.0.1
sphinx: 1.8.1
patsy: 0.5.1
dateutil: 2.7.3
pytz: 2018.5
blosc: None
bottleneck: None
tables: 3.4.4
numexpr: 2.6.8
feather: None
matplotlib: 3.0.1
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 1.0.1
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by GivyBoy**
- [x] closes #24845
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/indexes/category.py [hunk 1]
```diff
@@ -571,6 +571,27 @@ def map(self, mapper):
 
     def _concat(self, to_concat: list[Index], name: Hashable) -> Index:
         # if calling index is category, don't check dtype of others
+
+        permutations = True
```
Comment (jreback) [inline pandas/core/indexes/category.py:575]: this is going to be non performant

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/category.py (hunk 1 above)
Comment (GivyBoy) [reply pandas/core/indexes/category.py:577]: Hi @jreback! Thanks for the comment. I have made some adjustments, are they better? Thanks!

**Comment by GivyBoy** [comment]
Hi @mroeschke or @jreback, I would love to have some feedback on this PR. Thanks a lot!

File: doc/source/whatsnew/v1.5.0.rst [hunk 2]
```diff
@@ -435,6 +435,47 @@ Calling :meth:`.DataFrameGroupBy.value_counts` with ``observed=True`` would inco
     1  c    0
     dtype: int64
 
+.. _whatsnew_150.notable_bug_fixes.concat_unordered_categoricalindex_overrides_indices:
+
+Concatenating two ``CategoricalIndex`` variables with ``ordered=False`` yields correct result
```
Comment (mroeschke) [inline doc/source/whatsnew/v1.5.0.rst:440]: I don't think we need a whole entry for this bug fix. Just a 1 line description under the `Categorical` section.

Comment (GivyBoy) [reply doc/source/whatsnew/v1.5.0.rst]: Hi @mroeschke, thanks for the feedback. I updated the PR.

File: doc/source/whatsnew/v1.5.0.rst [hunk 3]
```diff
@@ -864,9 +864,13 @@ Bug fixes
 
 Categorical
 ^^^^^^^^^^^
+- Bug in :meth:`.Categorical.view` not accepting integer dtypes (:issue:`25464`)
+- Bug in :meth:`.CategoricalIndex.union` when the index's categories are integer-dtype and the index contains ``NaN`` values incorrectly raising instead of casting to ``float64`` (:issue:`45362`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v1.5.0.rst:868]: Looks like 2 entries got duplicated here.

File: doc/source/whatsnew/v1.5.0.rst [hunk 4]
```diff
@@ -864,9 +864,13 @@ Bug fixes
...
+- Bug in :meth:`DataFrame.concat` when concatenating two (or more) unordered ``CategoricalIndex`` variables, whose categories are permutations, yields incorrect index values (:issue:`24845`)
 - Bug in :meth:`Categorical.view` not accepting integer dtypes (:issue:`25464`)
 - Bug in :meth:`CategoricalIndex.union` when the index's categories are integer-dtype and the index contains ``NaN`` values incorrectly raising instead of casting to ``float64`` (:issue:`45362`)
 
+
```
Comment (mroeschke) [inline doc/source/whatsnew/v1.5.0.rst:873]: Could you remove these whitespaces? (And in the code below?)

Comment (GivyBoy) [reply doc/source/whatsnew/v1.5.0.rst:887]: Sure! Got them done

File: doc/source/whatsnew/v1.5.0.rst (hunk 3 above)
Comment (GivyBoy) [reply doc/source/whatsnew/v1.5.0.rst:885]: Strange. I think it's happening when I merge with the main branch.

Comment (GivyBoy) [reply doc/source/whatsnew/v1.5.0.rst:885]: Hi @mroeschke, could you point to the lines where the duplicates are? Thanks

File: doc/source/whatsnew/v1.5.0.rst [hunk 5]
```diff
@@ -882,8 +881,9 @@ Bug fixes
 
 Categorical
 ^^^^^^^^^^^
-- Bug in :meth:`Categorical.view` not accepting integer dtypes (:issue:`25464`)
-- Bug in :meth:`CategoricalIndex.union` when the index's categories are integer-dtype and the index contains ``NaN`` values incorrectly raising instead of casting to ``float64`` (:issue:`45362`)
+- Bug in :meth:`.Categorical.view` not accepting integer dtypes (:issue:`25464`)
+- Bug in :meth:`.CategoricalIndex.union` when the index's categories are integer-dtype and the index contains ``NaN`` values incorrectly raising instead of casting to ``float64`` (:issue:`45362`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v1.5.0.rst:885]:
```suggestion
```

File: doc/source/whatsnew/v1.5.0.rst (hunk 3 above)
Comment (GivyBoy) [reply doc/source/whatsnew/v1.5.0.rst]: Hi @mroeschke, I took your suggestions. Is there anything else in the code that needs work?

File: doc/source/whatsnew/v1.5.0.rst [hunk 6]
```diff
@@ -882,8 +881,7 @@ Bug fixes
 
 Categorical
 ^^^^^^^^^^^
-- Bug in :meth:`Categorical.view` not accepting integer dtypes (:issue:`25464`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v1.5.0.rst:885]: Oops, sorry it appears the _original_ entries got deleted too. Could you restore them? (the only change in this file should be 1 new entry

Comment (GivyBoy) [reply doc/source/whatsnew/v1.5.0.rst:885]: Yeah, I figured. That's why I was a bit confused. No worries!

**Comment by GivyBoy** [comment]
Hi @mroeschke, I have made the edits. I thought something was wrong with my commit, then I realized that a PR was merged with the same three errors. Could you explain, if possible, why this is happening? Thanks!

**Comment by mroeschke** [comment]
> Hi @mroeschke, I have made the edits. I thought something was wrong with my commit, then I realized that a PR was merged with the same three errors. Could you explain, if possible, why this is happening? Thanks!

We had some unrelated CI issues lately. Could you merge in main once more?

**Comment by GivyBoy** [comment]
Hi @mroeschke! I merged and I am receiving errors that I'm not receiving when I run the tests on my local machine.

**Comment by mroeschke** [comment]
> Hi @mroeschke! I merged and I am receiving errors that I'm not receiving when I run the tests on my local machine.

We experience a similar error on main that is affecting all builds. Merging one more time should fix it.

**Comment by GivyBoy** [comment]
It passed all test cases, @mroeschke. Thanks a lot for your guidance!

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks for sticking all the reviews this @GivyBoy!

**Pull request merged**